### DISCUSSION TITLE
Add force hard navigate option to router

### DISCUSSION THREAD

**Discussion by rijk** [vercel/next.js#44149] [Ideas]
### Describe the feature you'd like to request

It would be nice to have a `forceHardNavigation` option, akin to `forceOptimisticNavigation`, to opt out of soft navigation if you know the cache has been invalidated.

Use case:
﻿
1. Navigate to `/admin` → next-auth redirects to `/sign-in?callbackUrl=%2Fadmin`, router caches redirect
3. Sign in
4. Try to navigate user back with `router.replace(callbackUrl)` → Doesn't work because of the cached redirect, you are stuck on `/sign-in?callbackUrl=%2Fadmin`

### Describe the solution you'd like

Solution:

`router.replace(callbackUrl, {forceHardNavigation: true})`

Can be implemented [here](https://github.com/vercel/next.js/blob/canary/packages/next/client/components/reducer.ts#L851-L859) as follows:

```typescript
          const hardNavigate =
            // TODO-APP: Revisit if this is correct.
            search !== location.search ||
            action.forceHardNavigation || // <- newly added line
            shouldHardNavigate(
              // TODO-APP: remove ''
              ['', ...flightSegmentPath],
              state.tree,
              newTree
            )
```

### Describe alternatives you've considered

The only alternative right now is to either:

1. Do a hard reload through `window.location.href = '/admin'`
2. Add a `?_` location.search to force hard navigation → brittle as his behavior is undocumented and might change reading the todo comment. Also, this doesn't work if a link in your app navigates to `/admin`; the redirect is still cached so this will still redirect you to `/sign-in` (even though the user is already signed in).

**Comment by ayhanap** [comment]
I ended up doing `router.replace('/home'); router.refresh();`. Checked logs, and it does load the page a second time in the background though.

We definitely need this kind of utility.

**Comment by ianldgs** [comment]
I have the following situation:

1. Type "foo" in search form on the homepage
1. Navigate `/` -> `/search?q=foo`
1. Renders skeletons for multiple search providers (`Suspense`)
1. As each provider loads, the data shows (slow search providers don't block fast ones)

Then:
1. Type "bar" in search form on the search results page (do `router.push`)
1. Nothing happens for a while (this is due to one provider being slow)
1. Navigate `/search?q=foo` -> `/search?q=bar`
1. Show data for all search providers

Current workaround: using default `form[onSubmit]` behaviour and letting a regular browser navigation happen.

Maybe forcing a hard navigation here would help.

**Reply by JakeElder** [reply]
This is still the case.. It seems like a serious oversight. Are there any other solutions?

**Reply by ianldgs** [reply]
Yeah, seems like there was a regression.
I don't know of any other solutions, unfortunately :/

**Comment by rijk** [comment]
Sounds like this is done in https://github.com/vercel/next.js/pull/46723

**Reply by ianldgs** [reply]
Indeed, I just tried 13.3.0 and the problem I described no longer happens! Thanks a lot for the heads-up.

**Reply by rijk** [reply]
Yep, just tested in my app and this now works flawlessly:

```ts
// after sign in
router.refresh()
router.replace(callbackUrl)
```
